### DISCUSSION TITLE
Feedback: Script `worker` strategy

### DISCUSSION THREAD

**Discussion by housseindjirdeh** [vercel/next.js#36778] [Show and tell]
*tl.dr:* Thread to collect feedback on `<Script strategy="worker" />`

----

Details: In v12.1.1-canary.11, `next/script` introduced a new `worker` strategy that allows you to relocate and execute third-party scripts in a web worker with Partytown (see [documentation](https://nextjs.org/docs/basic-features/script#off-loading-scripts-to-a-web-worker-experimental)). This strategy is still experimental but I would like to use this thread to gather all types of feedback.

Some questions that may be useful to answer (although any feedback is worth sharing):

- Have you noticed any performance benefits after switching to the `worker` strategy for any scripts on your site?
- Have there been any scripts that failed to load after you switched to using the `worker` strategy?
- Do you have any problems with the current developer experience?
- Is there anything that can be improved with the current documentation?

**Comment by mitestainer** [comment]
We've been testing the `worker` approach in our websites here at [Apto](https://apto.vc/) internally and the results have been positive. While running some page speed/web vitals audits we could notice improvements related to the impact of third-party code, unused javascript code and the main thread work time. That's what we wanted to check, and it worked alright!

We got the CORS error every time, even after applying the forward configuration Partytown suggests in their docs. We proxied the requests, but the Bing Search snippet coming from GTM, when present, started to throw an `UET is not a constructor` error (I've seen people reporting this at Partytown's repo). There's also no integration with the GTM test tool (also already reported to Partytown). Due to that, we decided to hold the implementation of `worker` and keep on doing more tests as more updates on Next.js and Partytown rollout.

The idea behind Partytown is really interesting and I'm glad you guys are working on integrate that on Next.js. Thanks for the great job! We'll be looking forward to the updates and we hope to see it working in our websites soon.

**Reply by housseindjirdeh** [reply]
Thanks for the feedback! I'm glad to hear that early testing showed positive results. Do you happen to have any numbers that you can share with regards to how much it improved main thread work time or any of the other metrics? 

I totally understand why you'd like to hold off fully integrating it to your sites until those issues get resolved. I'll send this feedback over to the Partytown folks and hopefully they can get resolved in the near future 🙏

**Reply by mitestainer** [reply]
Sure! We ran several web.dev and WebPageTest audits to come to our conclusion. All the code related to third-party scripts didn't show on the _unused javascript_ tab, we managed to decreased the _main-thread work_ metric from 34.4s to 9.4s and the main-thread blocking due to third-prty code went down to zero.

**Comment by axeldelafosse** [comment]
Hey @housseindjirdeh! Thank you so much for your work, this is something I want to do since a long time now!

Here is a PR where I'm loading RudderStack with `next/script` using the `worker` strategy instead of using the npm package: https://github.com/showtime-xyz/showtime-frontend/pull/1383

Unfortunately the script isn't loading. I'm getting this error:

- Safari
<img width="1350" alt="Screenshot 2022-08-10 at 16 07 53" src="https://user-images.githubusercontent.com/10477267/183931440-506a9f1f-79c4-4601-a4c9-58a67c43dfe9.png">
<img width="1108" alt="Screenshot 2022-08-10 at 16 04 13" src="https://user-images.githubusercontent.com/10477267/183931451-d81ced9c-21e6-4928-b1b4-0da695d070ad.png">

- Brave
<img width="662" alt="Screenshot 2022-08-10 at 16 39 34" src="https://user-images.githubusercontent.com/10477267/183931511-3f5bb3e9-3c1c-4e9e-bcd6-3f93dc1c4c31.png">

Any idea? Have you heard of someone successfully using Partytown to load RudderStack?

**Comment by hieunguyen1704** [comment]
Hi @housseindjirdeh, thank you for your great efforts.
But when I try with this, I've got some issues on Chrome, Safari still works fine:

Specifically, on the first load of the page everything works fine, but on the 2nd reload I get a 404 Not found error sending a request to get file partytown-sandbox-sw.html (http:// localhost:3000/_next/static/~partytown/debug/partytown-sandbox-sw.html?1665936200847). In Partytown's documentation they mentioned that if this request fails, then service worker is not installed correctly:

![image](https://user-images.githubusercontent.com/44908179/196047636-004f9e46-60b8-445a-8d05-c85f02a44573.png)

On Chrome:
![image](https://user-images.githubusercontent.com/44908179/196047719-4092b014-ffd8-4b18-8de7-7f4993c3b479.png)

On Safari:
![image](https://user-images.githubusercontent.com/44908179/196047791-87698e4e-5bfd-4f30-a19f-5945512e0234.png)

Here is my `_document.js` :

```
import { Html, Head, Main, NextScript } from "next/document";
import Script from "next/script";

export default function Document() {
  return (
    <Html>
      <Head>
        <script
          data-partytown-config
          dangerouslySetInnerHTML={{
            __html: `
              partytown = {
                lib: "/_next/static/~partytown/",
                debug: true
              };
            `,
          }}
        />
      </Head>
      <body>
        <Main />
        <NextScript />
        <Script
          id="gtag-init"
          strategy="worker"
          dangerouslySetInnerHTML={{
            __html: `!function(){function t(){var t,e,n,a;t=window,e=document,n="script",t[a="dataLayer"]=t[a]||[],t[a].push({"gtm.start":(new Date).getTime(),event:"gtm.js"}),a=e.getElementsByTagName(n)[0],(n=e.createElement(n)).async=!0,n.src="https://www.googletagmanager.com/gtm.js?id=GTM-5D3JJZT",a.parentNode.insertBefore(n,a)}function e(){setTimeout(t,0)}window.addEventListener?window.addEventListener("load",e,!1):window.attachEvent&&window.attachEvent("onload",e)}();`,
          }}
        />
      </body>
    </Html>
  );
}
```
Next version: 12.3.1
@builder.io/partytown version: 0.7.1
